enhancements/installer/azure-support-known-cloud-environments.md

```diff
@@ -0,0 +1,231 @@
+
+The controller will use the `cloudName` from the Infrastructure object to set the [`cloud`](https://github.com/kubernetes/kubernetes/blob/89ba90573f163ee3452b526f30348a035d54e870/staging/src/k8s.io/legacy-cloud-providers/azure/auth/azure_auth.go#L45-L46) field for the Azure cloud config.
+
+If the user has already set a `cloud` value in the `.spec.cloudConfig` which doesn't match the value provided by the Infrastructure object, the controller should return error detailing the conflict in values.
+
+#### Installer
+
+##### Install Config
+
+The install config should allow the user to set the cloud name for Azure.
+
+```go
+// types/pkg/azure/platform.go
+type Platform struct {
+    ...
+	// cloudName is the name of the Azure cloud environment which can be used to configure the Azure SDK
+	// with the appropiate Azure API endpoints.
+	// If empty, the value is equal to `AzurePublicCloud`.
+	// +optional
+	CloudName AzureCloudEnvironment `json:"cloudName,omitempty"`
+}
+
+// +kubebuilder:validation:Enum="";AzurePublicCloud;AzureUSGovernmentCloud;AzureChinaCloud;AzureGermanCloud
+type AzureCloudEnvironment string
+
+const (
+	PublicAzureCloud AzureCloudEnvironment = "AzurePublicCloud"
+	USGovernmentAzureCloud AzureCloudEnvironment = "AzureUSGovernmentCloud"
+	AzureChinaCloud AzureCloudEnvironment = "AzureChinaCloud"
+	AzureGermanCloud AzureCloudEnvironment = "AzureGermanCloud"
+)
+```
+
+##### Configuring the Azure SDK
+
+The installer uses the Authorizers generated from [`newSessionFromFile`](https://github.com/openshift/installer/blob/a11cc4d29231135cd889487275ee32a20546b7e2/pkg/asset/installconfig/azure/session.go#L50) to create clients for Azure APIs. Updating the function to accept Azure cloud environment name should allow correct initiation of the clients.
+
+##### Configuring the internal terraform-provider-azurerm with environment
+
+The AzureRM provider accepts [`environment`](https://www.terraform.io/docs/providers/azurerm/index.html#environment) field to configure the Azure SDK. The valid values are `public`, `usgovernment`, `german`, and `china`, the installer should convert the install-config.yaml `AzureCloudEnvironment` string to a value acceptable by the provider.
+
+##### Configuring infrastructure object
+
+The installer should transform the install-config.yaml `AzureCloudEnvironment` string to set the `.status.platformStatus.azure.cloudName` in the Infrastructure object as detailed [above](#Infrastructure-global-configuration)
+
+##### Destroy
+
+The metadata.json needs to store the cloud name from the install-config.yaml so that the users can delete the clusters without any previous state.
+
+#### Cluster Operators
+
+Almost of the cluster operators will have to read the cloud name from the Infrastructure's object's status for Azure and make sure the operators are watching for changes.
```

Comment:
Can the cluster really change during runtime from azurepublic to azuregov?

Reply:
> Can the cluster really change during runtime from azurepublic to azuregov?

No, as described earlier in the doc:

> The Azure cloud is a one-time setting that the user shouldn't be allowed to change and therefore, this information will only be allowed in the `status` section.

So I don't see a problem with consumers that only poll once on startup.  Should we point out that both approaches (read-once and watching) are acceptable?

Reply:
For now i have dropped the requirement to always sync, as that is not yet required to good to have because, there is near term goal to allow users to change the _individual_ endpoints which would require ability to act on change of cloud name to say govcloud to customcloud

Reply:
done